### PR TITLE
Pub: Add support for path dependencies

### PR DESCRIPTION
**Pull request by bennati**
Signed-off-by: Stefano Bennati <stefano.bennati@here.com>

Please ensure that your pull request adheres to our [contribution guidelines](https://github.com/oss-review-toolkit/.github/blob/main/CONTRIBUTING.md). Thank you!

### CODE REVIEW

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Pub.kt:117]: Please try to avoid creating such invalid empty file objects.

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt:84]: Wrong indentation (also see failing `detekt` checks).

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt:83]: Both `return` statements should be replaced by a single `return` before the `if`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt:117]: Can you please suggest a correction? I'm new to Kotlin. Thanks!

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Pub.kt:117]: That's actually not a Kotlin-thing, you wouldn't do this in Java either. Also using `var`s is discouraged in Kotlin in favor of using `val`s where possible. It seems to me you only "need" this because you're not passing the `workingDir` as a parameter to all functions that need it. So I guess you could change the function signatures accordingly instead, so the `workingDir` is available everywhere you need it.

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt]: done

**Review by sschuberth — CHANGES_REQUESTED** [review]
I'm done with the formal aspects, but would like @MarcelBochtler to comment on the technical aspects.

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: removed

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt:458]: done

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt]: done

File: analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt
Comment (bennati) [reply analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt]: done

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (MarcelBochtler) [inline analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt:78]: While you tried to keep the wording similar to the other comments of this if, I find this comment quite confusing.
How about:
```
For "path" packages, the path should be the absolut resolved path of the "path" given in the description.
```

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (MarcelBochtler) [inline analyzer/src/main/kotlin/managers/Pub.kt:477]: I'm not sure about this. Having only the `path` in the `VcsInfo` might not be correct, as the "path" dependency might point to a location on the file system not managed via VCS.
Is the Downloader able to "download" sources for a package with a `VcsInfo` like this?

**Review by MarcelBochtler — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt:76]: I have created a separate `if` with a separate return because the return statement at line 104 makes operations on `path` that are not compatible with the operations I need to perform in case `path` is a local directory.

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: yes sorry, now done

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: done

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt]: done

**Comment by sschuberth** [comment]
@bennati, please remember to click on the "Re-request review" icon in the upper-right "Reviewers" section if you're finished with addressing review comments and would like to get another review.

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Pub.kt:459]: Broken indentation.

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Pub.kt:474]: > Is the Downloader able to "download" sources for a package with a `VcsInfo` like this?

No 😀 

My proposal would be to
1. only support the case when the package directory is below the project directory, otherwise log a warning
2. derive the `VcsInfo` from the project like: `VersionControlSystem.forDirectory(workingDir).getInfo()`
3. amend that VcsInfo with the path from `pkgInfoFromLockFile["description"]["path"]` (relative to the project root)

@bennati would that cover your use-case? I.e. where are the paths to the dependencies located in your case?

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (MarcelBochtler) [inline analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt:77]: This line is too long. That's why detekt is complaining. You can run detect locally using `./gradlew detekt`.

**Review by MarcelBochtler — DISMISSED** [review]

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: done

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt]: done

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt]: done

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Pub.kt:474]: > What do you mean with 3? the path should be relative to the project directory, instead of absolute?

Yes. Or phrased differently, a `VcsInfo`'s path must be relative to the repository root. Kotlin has a `relativeTo` extension function on `File`s that might come handy here.

> the path in a VcsInfo cannot be amended because it is a `val`, or am I doing something wrong?

The way to do this is via a `data class`' built-in `copy` function: `vcsInfo.copy(path = pkgInfoFromLockFile["description"]["path"])`.

Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: done

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt:74]: I had to undo this because using `.boleeanValue()` would return a NullPointerException any time the 'relative' key is not defined in the description.

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/utils/PubCacheReader.kt:74]: Then `description["relative"]?.booleanValue() ?: false` should do the trick.

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: done

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: done

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: Your suggestion is based on outdated code, I adapted it to the new code

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: done

File: analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt
Comment (bennati) [reply analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt]: "relative" refers to the test using a relative path, I renamed the variables to clarify this

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (sschuberth) [inline analyzer/src/main/kotlin/managers/Pub.kt:471]: Nits: Please single-quite `'$path'` and `'$workingDir'` in case they contain spaces, and end the log sentence with a dot.

**Review by sschuberth — COMMENTED** [review]
This LGTM now (despite my nits), but I'd like also @MarcelBochtler to approve.

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Pub.kt:471]: And @bennati please also look into the *detekt* issues about too long lines.

Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: done

Comment (sschuberth) [reply analyzer/src/main/kotlin/managers/Pub.kt]: There now seems to be [some breakage](https://github.com/oss-review-toolkit/ort/runs/6421480451?check_suite_focus=true#step:4:256).

Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: The issue with `workingDir` not being passed is related to this https://github.com/oss-review-toolkit/ort/pull/5330#discussion_r868175788
How can I mark it as an optional parameter?

File: analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt
Comment (sschuberth) [inline analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt:105]: You also need to add this line to the existing test cases above now.

Comment (bennati) [reply analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt]: done

File: analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt
Comment (bennati) [reply analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt]: done

File: analyzer/src/main/kotlin/managers/Pub.kt
Comment (bennati) [reply analyzer/src/main/kotlin/managers/Pub.kt]: Kept the parameter as is and added it to the calls in the other tests

File: analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt
Comment (sschuberth) [inline analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt:35]: This should become `const` to silence he *detekt* issue.

Comment (sschuberth) [reply analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt:35]: And while you're it, maybe you could also fix a commit message nit, and write capitalized "Add" after the colon in the title.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by MarcelBochtler — COMMENTED** [review]
LGTM if the last `detekt` issue is solved.

File: analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt
Comment (bennati) [reply analyzer/src/test/kotlin/managers/utils/PubCacheReaderTest.kt:35]: done

**Comment by bennati** [comment]
While this code correctly recognizes path-based dependencies, I noticed that the scanner now detects these dependencies twice: once with the name provided in the `pubspec.yaml` file and another with the directory name. 

For example, if the pubspec contains:
```
dependencies:
  name:
    path: name1
```

The `analyzer-result` will contain only one entry:
```
{
"id" : "Pub:name:name:",
"vcs_processed": {
  "path" : "name1"
}
```

While the `scan-result` will contain two entries:

```
{
"id" : "Pub:name:name:",
"provenance": {
  "vcs_info" : {
    "path" : "name1"
  }
},
...
{
"id" : "Pub:name1:name1:",
"provenance": {
  "vcs_info" : {
    "path" : "name1"
  }
},
```

Any idea how to solve this?

**Comment by sschuberth** [comment]
> Any idea how to solve this?

I believe what you see is https://github.com/oss-review-toolkit/ort/issues/5365 and thus nothing that can be fixed as part of your PR.

**Review by sschuberth — APPROVED** [review]

**Pull request merged**
